src/pages/Dashboard.tsx

```diff
@@ -43,101 +43,148 @@ const Dashboard: React.FC = () => {
   const [searchTerm, setSearchTerm] = useState('');
   const { isOpen, onToggle } = useDisclosure();
   const { colorMode } = useColorMode();
+  const [sortConfig, setSortConfig] = useState<{
+    key: keyof IData;
+    direction: 'asc' | 'desc';
+  }>({
+    key: 'lp',
+    direction: 'asc',
+  });
 
   const bgColor = colorMode === 'dark' ? 'gray.800' : 'white';
   const borderColor = colorMode === 'dark' ? 'gray.700' : 'gray.200';
 
-  const mappedData = countriesData?.map((country, index) => ({
-    lp: index,
-    code: country.cca3,
-    name: country.name.common,
-    flag: country.flags?.png,
-    capital: country.capital.join(', '),
-    population: country.population,
-    independent: country.independent,
-    area: country.area,
-  })) as IData[];
+  const filteredData = useMemo(() => {
+    if (!countriesData) return [];
+    const mappedData = countriesData.map((country, index) => ({
+      lp: index,
+      code: country.cca3,
+      name: country.name.common,
+      flag: country.flags?.png,
+      capital: country.capital.join(', '),
+      population: country.population,
+      independent: country.independent,
+      area: country.area,
+    })) as IData[];
+
+    const filtered = mappedData.filter(
+      (item) =>
+        !searchTerm ||
+        item.name.toLowerCase().includes(searchTerm.toLowerCase())
+    );
+
+    filtered.sort((a, b) => {
+      if (a[sortConfig.key] < b[sortConfig.key]) {
+        return sortConfig.direction === 'asc' ? -1 : 1;
+      }
+      if (a[sortConfig.key] > b[sortConfig.key]) {
+        return sortConfig.direction === 'asc' ? 1 : -1;
+      }
+      return 0;
+    });
+
+    return filtered;
+  }, [countriesData, searchTerm, sortConfig]);
+
+  const handleSortChange = (
+    ev: React.MouseEvent<HTMLElement>,
+    column: TableColumn
+  ) => {
+    const key = column.fieldName as keyof IData;
+    setSortConfig((prevState) => ({
+      key,
+      direction:
+        prevState.key === key && prevState.direction === 'asc' ? 'desc' : 'asc',
+    }));
+  };
 
   const columns: TableColumn[] = [
     {
       key: 'lp',
       fieldName: 'lp',
       name: 'LP',
+      onColumnClick: handleSortChange,
+      isSorted: sortConfig.key === 'lp',
+      isSortedDescending: sortConfig.direction === 'desc',
     },
     {
       key: 'name',
       fieldName: 'name',
       name: 'Name',
+      onColumnClick: handleSortChange,
+      isSorted: sortConfig.key === 'name',
+      isSortedDescending: sortConfig.direction === 'desc',
     },
     {
       key: 'flag',
       fieldName: 'flag',
       name: 'Flag',
-      onRender(item: IData) {
-        return <Image src={item.flag} alt="Flag" height="20px" />;
-      },
+      onRender: (item: IData) => (
+        <Image src={item.flag} alt="Flag" height="20px" />
+      ),
     },
     {
       key: 'capital',
       fieldName: 'capital',
       name: 'Capital',
+      onColumnClick: handleSortChange,
+      isSorted: sortConfig.key === 'capital',
```

Comment:
Tutaj wystarczy wpisać wartości "true", w przypadku gdy chcesz żeby na initcie sortowało według tej kolumny. Przez co unikniesz wpisywania tych warunków przy każdym z kluczy.
`{
key: 'capital', 
fieldName: 'capital',
name: 'Capital',
onColumnClick: handleSortChange,
isSorted: true,
isSortedDescending: true
}`

Wtedy tą zmienna "columns" możesz wrzucić do jakiegoś state-a. 

I na initcie sprawdzić, który z elementów ma ustawione sortowanie i na podstawie tego przesortować "data".

I w przypadku gdy ktoś kliknie na kolumnę to aktualizujesz ten state z kolumnami.